omegaconf/dictconfig.py

```diff
@@ -176,6 +176,10 @@ def _validate_set_merge_impl(self, key: Any, value: Any, is_assign: bool) -> Non
         if value == "???":
             return
 
+        if is_assign and isinstance(value, ValueNode) and self._has_element_type():
+            self._check_assign_value_node(key, value)
```

Comment:
Looks like you modeled your check on it, but you did it wrong and now the error you copied does not match your check.
try to consolidate this logic _validate_set_merge_impl().
along the way, see if you can do the TODO in it and simplify it a bit.